File: sdk/flutter/gizclaw/lib/src/signaling.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'transport.dart';

const _base58Alphabet =
    '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
final _base58Map = <int, int>{
  for (var i = 0; i < _base58Alphabet.length; i++)
    _base58Alphabet.codeUnitAt(i): i,
};

class GiznetSignalingIdentity {
  const GiznetSignalingIdentity({
    this.clientPublicKey,
    required this.clientPrivateKey,
    required this.serverPublicKey,
  });

  final List<int> clientPrivateKey;
  final List<int>? clientPublicKey;
  final List<int> serverPublicKey;
}

class PreparedGiznetWebRtcOffer {
  const PreparedGiznetWebRtcOffer({
    required this.body,
    required this.clientPublicKey,
    required this.nonce,
    required this.openAnswer,
    required this.timestamp,
  });

  final Uint8List body;
  final String clientPublicKey;
  final String nonce;
  final Future<String> Function(List<int> encryptedAnswer) openAnswer;
  final int timestamp;
}

class GiznetServerInfo {
  const GiznetServerInfo({
    this.endpoint,
    this.protocol,
    required this.publicKey,
    this.signalingPath = giznetWebRtcSignalingPath,
  });

  final String? endpoint;
  final String? protocol;
  final String publicKey;
  final String signalingPath;

  factory GiznetServerInfo.fromJson(Map<String, Object?> json) {
    final protocol = json['protocol'] as String?;
    if (protocol != null && protocol != 'gizclaw-webrtc') {
      throw FormatException(
        'server-info protocol = $protocol, want gizclaw-webrtc',
      );
    }
    final publicKey = (json['public_key'] as String?)?.trim();
    if (publicKey == null || publicKey.isEmpty) {
      throw const FormatException('server-info missing public_key');
    }
    final publicKeyBytes = base58Decode(publicKey);
    if (publicKeyBytes.length != 32 ||
        publicKeyBytes.every((byte) => byte == 0)) {
      throw const FormatException('server-info invalid public_key');
    }
    final signalingPath = _normalizeSignalingPath(
      json['signaling_path'] as String?,
    );
    return GiznetServerInfo(
      endpoint: json['endpoint'] as String?,
      protocol: protocol,
      publicKey: publicKey,
      signalingPath: signalingPath,
    );
  }
}

Future<PreparedGiznetWebRtcOffer> prepareEncryptedGiznetWebRtcOffer(
  GiznetSignalingIdentity identity,
  String offerSdp, {
  List<int>? nonceBytes,
  int? timestamp,
}) async {
  final clientPrivateKey = _expectKeyBytes(
    identity.clientPrivateKey,
    'client private key',
  );
  final x25519 = X25519();
  final clientKeyPair = await x25519.newKeyPairFromSeed(clientPrivateKey);
  final derivedPublicKey = await clientKeyPair.extractPublicKey();
  final clientPublicKey = identity.clientPublicKey == null
      ? derivedPublicKey.bytes
      : _expectKeyBytes(identity.clientPublicKey!, 'client public key');
  final serverPublicKey = _expectKeyBytes(
    identity.serverPublicKey,
    'server public key',
  );
  final nonce = base64UrlEncodeNoPadding(
    nonceBytes == null ? randomBytes(16) : Uint8List.fromList(nonceBytes),
  );
  final timestampValue =
      timestamp ?? DateTime.now().millisecondsSinceEpoch ~/ 1000;
  final keys = await _deriveSignalingKeys(
    clientKeyPair: clientKeyPair,
    serverPublicKey: serverPublicKey,
    nonce: nonce,
    timestamp: timestampValue,
  );
  final requestAad = signalingAad(clientPublicKey, timestampValue, nonce);
  final cipher = Chacha20.poly1305Aead();
  final encrypted = await cipher.encrypt(
    utf8.encode(offerSdp),
    secretKey: SecretKey(keys.requestKey),
    nonce: keys.requestNonce,
    aad: requestAad,
  );

  return PreparedGiznetWebRtcOffer(
    body: encrypted.concatenation(nonce: false),
    clientPublicKey: base58Encode(clientPublicKey),
    nonce: nonce,
    openAnswer: (encryptedAnswer) async {
      if (encryptedAnswer.length < 16) {
        throw const FormatException(
          'encrypted answer is shorter than AEAD tag',
        );
      }
      final responseAad = signalingAad(
        clientPublicKey,
        timestampValue,
        nonce,
        answer: true,
      );
      final secretBox = SecretBox(
        encryptedAnswer.sublist(0, encryptedAnswer.length - 16),
        nonce: keys.responseNonce,
        mac: Mac(encryptedAnswer.sublist(encryptedAnswer.length - 16)),
      );
      final bytes = await cipher.decrypt(
        secretBox,
        secretKey: SecretKey(keys.responseKey),
        aad: responseAad,
      );
      return utf8.decode(bytes);
    },
    timestamp: timestampValue,
  );
}

Uint8List signalingAad(
  List<int> clientPublicKey,
  int timestamp,
  String nonce, {
  bool answer = false,
}) {
  final parts = [
    'POST',
    giznetWebRtcSignalingPath,
    base58Encode(clientPublicKey),
    timestamp.toString(),
    nonce,
    if (answer) 'answer',
  ];
  return Uint8List.fromList(utf8.encode(parts.join('\n')));
}

String base58Encode(List<int> bytes) {
  var value = BigInt.zero;
  for (final byte in bytes) {
    value = (value << 8) + BigInt.from(byte);
  }
  var text = '';
  while (value > BigInt.zero) {
    final mod = (value % BigInt.from(58)).toInt();
    text = _base58Alphabet[mod] + text;
    value ~/= BigInt.from(58);
  }
  for (final byte in bytes) {
    if (byte != 0) {
      break;
    }
    text = '1$text';
  }
  return text.isEmpty ? '1' : text;
}

Uint8List base58Decode(String text) {
  var value = BigInt.zero;
  for (final unit in text.codeUnits) {
    final digit = _base58Map[unit];
    if (digit == null) {
      throw FormatException(
        'invalid base58 character ${String.fromCharCode(unit)}',
      );
    }
    value = value * BigInt.from(58) + BigInt.from(digit);
  }
  final bytes = <int>[];
  while (value > BigInt.zero) {
    bytes.add((value & BigInt.from(0xff)).toInt());
    value >>= 8;
  }
  for (final unit in text.codeUnits) {
    if (unit != '1'.codeUnitAt(0)) {
      break;
    }
    bytes.add(0);
  }
  return Uint8List.fromList(bytes.reversed.toList());
}

String base64UrlEncodeNoPadding(List<int> bytes) =>
    base64Url.encode(bytes).replaceAll('=', '');

Uint8List base64UrlDecodeNoPadding(String text) {
  final normalized = text.padRight((text.length + 3) ~/ 4 * 4, '=');
  return Uint8List.fromList(base64Url.decode(normalized));
}

Uint8List randomBytes(int length, {Random? random}) {
  final source = random ?? Random.secure();
  return Uint8List.fromList(
    List<int>.generate(length, (_) => source.nextInt(256)),
  );
}

String _normalizeSignalingPath(String? path) {
  final value = path?.trim() ?? '';
  if (value.isEmpty) {
    return giznetWebRtcSignalingPath;
  }
  if (!value.startsWith('/') || value.startsWith('//')) {
    throw FormatException('server-info invalid signaling_path $value');
  }
  return value;
}

Uint8List _expectKeyBytes(List<int> bytes, String name) {
  if (bytes.length != 32) {
    throw ArgumentError.value(bytes.length, name, 'invalid key length');
  }
  return Uint8List.fromList(bytes);
}

Future<_SignalingKeys> _deriveSignalingKeys({
  required SimpleKeyPair clientKeyPair,
  required List<int> serverPublicKey,
  required String nonce,
  required int timestamp,
}) async {
  final x25519 = X25519();
  final shared = await x25519.sharedSecretKey(
    keyPair: clientKeyPair,
    remotePublicKey: SimplePublicKey(serverPublicKey, type: KeyPairType.x25519),
  );
  final salt = Uint8List.fromList([
    ...base64UrlDecodeNoPadding(nonce),
    ...utf8.encode(timestamp.toString()),
  ]);
  final hkdf32 = Hkdf(hmac: Hmac.sha256(), outputLength: 32);
  final hkdf12 = Hkdf(hmac: Hmac.sha256(), outputLength: 12);
  final requestKey = await hkdf32.deriveKey(
    secretKey: shared,
    nonce: salt,
    info: utf8.encode('giznet/gizwebrtc/http-signaling/v1 c2s'),
  );
  final requestNonce = await hkdf12.deriveKey(
    secretKey: shared,
    nonce: salt,
    info: utf8.encode('giznet/gizwebrtc/http-signaling/v1 c2s nonce'),
  );
  final responseKey = await hkdf32.deriveKey(
    secretKey: shared,
    nonce: salt,
    info: utf8.encode('giznet/gizwebrtc/http-signaling/v1 s2c'),
  );
  final responseNonce = await hkdf12.deriveKey(
    secretKey: shared,
    nonce: salt,
    info: utf8.encode('giznet/gizwebrtc/http-signaling/v1 s2c nonce'),
  );
  return _SignalingKeys(
    requestKey: requestKey.bytes,
    requestNonce: requestNonce.bytes,
    responseKey: responseKey.bytes,
    responseNonce: responseNonce.bytes,
  );
}

class _SignalingKeys {
  const _SignalingKeys({
    required this.requestKey,
    required this.requestNonce,
    required this.responseKey,
    required this.responseNonce,
  });

  final List<int> requestKey;
  final List<int> requestNonce;
  final List<int> responseKey;
  final List<int> responseNonce;
}

File: sdk/js/gizclaw/signaling.ts
import { create, toBinary } from "@bufbuild/protobuf";
import {
  AdmissionCredentialSchema,
  type AdmissionCredential as AdmissionCredentialMessage,
} from "./generated/giznet/admission_pb.js";

import { chacha20poly1305 } from "@noble/ciphers/chacha.js";
import { x25519 } from "@noble/curves/ed25519.js";
import { hkdf } from "@noble/hashes/hkdf.js";
import { sha256 } from "@noble/hashes/sha2.js";

/** Structured Giznet credential; only the receiving policy interprets fields. */
export type AdmissionCredential = Pick<
  AdmissionCredentialMessage,
  "version" | "type" | "value"
>;

/** Maximum encoded protobuf credential bytes inside an encrypted offer. */
export const GIZNET_MAX_CREDENTIAL_BYTES = 4096;
/** Maximum UTF-8 bytes in the opaque credential value. */
export const GIZNET_MAX_CREDENTIAL_VALUE_BYTES = 512;
/** Leaves room within the 256 KiB signaling body for credentials and AEAD. */
export const GIZNET_MAX_OFFER_SDP_BYTES =
  256 * 1024 - GIZNET_MAX_CREDENTIAL_BYTES - 7 - 16;

const signalingPath = "/webrtc/v1/offer";
const base58Alphabet =
  "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
const base58Map = new Map(
  [...base58Alphabet].map((char, index) => [char, index]),
);

export type GiznetSignalingIdentity = {
  clientPrivateKey: Uint8Array;
  clientPublicKey?: Uint8Array | string;
  serverPublicKey: Uint8Array | string;
};

export type PreparedGiznetWebRTCOffer = {
  body: Blob | File;
  clientPublicKey: string;
  nonce: string;
  openAnswer: (encryptedAnswer: Blob) => Promise<string>;
  timestamp: number;
};

export async function prepareEncryptedGiznetWebRTCOffer(
  identity: GiznetSignalingIdentity,
  offerSDP: string,
  credential?: AdmissionCredential,
): Promise<PreparedGiznetWebRTCOffer> {
  if (offerSDP.length > GIZNET_MAX_OFFER_SDP_BYTES)
    throw new Error("offer SDP exceeds signaling limit");
  const sdp = new TextEncoder().encode(offerSDP);
  if (sdp.byteLength > GIZNET_MAX_OFFER_SDP_BYTES)
    throw new Error("offer SDP exceeds signaling limit");
  const encoded = encodeAdmissionCredential(credential);
  let plaintext: Uint8Array = sdp;
  if (encoded != null) {
    plaintext = concatBytes([
      new Uint8Array([
        0x47,
        0x5a,
        0x4f,
        0x46,
        1,
        encoded.byteLength >> 8,
        encoded.byteLength & 0xff,
      ]),
      encoded,
      sdp,
    ]);
  }
  const clientPrivateKey = expectKeyBytes(
    identity.clientPrivateKey,
    "client private key",
  );
  const clientPublicKey =
    typeof identity.clientPublicKey === "string"
      ? base58Decode(identity.clientPublicKey)
      : identity.clientPublicKey != null
        ? expectKeyBytes(identity.clientPublicKey, "client public key")
        : x25519.getPublicKey(clientPrivateKey);
  const serverPublicKey =
    typeof identity.serverPublicKey === "string"
      ? base58Decode(identity.serverPublicKey)
      : expectKeyBytes(identity.serverPublicKey, "server public key");
  const nonce = randomNonce();
  const timestamp = Math.floor(Date.now() / 1000);
  const keys = deriveSignalingKeys(
    clientPrivateKey,
    serverPublicKey,
    nonce,
    timestamp,
  );
  const requestAAD = signalingAAD(clientPublicKey, timestamp, nonce, false);
  const body = chacha20poly1305(
    keys.requestKey,
    keys.requestNonce,
    requestAAD,
  ).encrypt(plaintext);

  return {
    body: new Blob([arrayBufferFromBytes(body)]),
    clientPublicKey: base58Encode(clientPublicKey),
    nonce,
    openAnswer: async (encryptedAnswer: Blob) => {
      const encrypted = new Uint8Array(await encryptedAnswer.arrayBuffer());
      const responseAAD = signalingAAD(clientPublicKey, timestamp, nonce, true);
      const answer = chacha20poly1305(
        keys.responseKey,
        keys.responseNonce,
        responseAAD,
      ).decrypt(encrypted);
      return new TextDecoder().decode(answer);
    },
    timestamp,
  };
}

/** Encodes and bounds a credential before discovery, ICE or signaling I/O. */
export function encodeAdmissionCredential(
  credential?: AdmissionCredential,
): Uint8Array | undefined {
  if (credential == null) return undefined;
  const utf8 = new TextEncoder();
  if (
    utf8.encode(credential.type).length > 128 ||
    utf8.encode(credential.value).length > GIZNET_MAX_CREDENTIAL_VALUE_BYTES
  ) {
    throw new Error("invalid admission credential length");
  }
  const encoded = toBinary(
    AdmissionCredentialSchema,
    create(AdmissionCredentialSchema, credential),
  );
  if (encoded.length === 0 || encoded.length > GIZNET_MAX_CREDENTIAL_BYTES) {
    throw new Error("invalid admission credential length");
  }
  return encoded;
}

function deriveSignalingKeys(
  clientPrivateKey: Uint8Array,
  serverPublicKey: Uint8Array,
  nonce: string,
  timestamp: number,
) {
  const shared = x25519.getSharedSecret(clientPrivateKey, serverPublicKey);
  const salt = concatBytes([
    base64URLDecode(nonce),
    new TextEncoder().encode(String(timestamp)),
  ]);
  return {
    requestKey: hkdf(
      sha256,
      shared,
      salt,
      new TextEncoder().encode("giznet/gizwebrtc/http-signaling/v1 c2s"),
      32,
    ),
    requestNonce: hkdf(
      sha256,
      shared,
      salt,
      new TextEncoder().encode("giznet/gizwebrtc/http-signaling/v1 c2s nonce"),
      12,
    ),
    responseKey: hkdf(
      sha256,
      shared,
      salt,
      new TextEncoder().encode("giznet/gizwebrtc/http-signaling/v1 s2c"),
      32,
    ),
    responseNonce: hkdf(
      sha256,
      shared,
      salt,
      new TextEncoder().encode("giznet/gizwebrtc/http-signaling/v1 s2c nonce"),
      12,
    ),
  };
}

function signalingAAD(
  clientPublicKey: Uint8Array,
  timestamp: number,
  nonce: string,
  answer: boolean,
): Uint8Array {
  const parts = [
    "POST",
    signalingPath,
    base58Encode(clientPublicKey),
    String(timestamp),
    nonce,
  ];
  if (answer) {
    parts.push("answer");
  }
  return new TextEncoder().encode(parts.join("\n"));
}

function randomNonce(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return base64URLEncode(bytes);
}

export function base58Decode(text: string): Uint8Array {
  let value = 0n;
  for (const char of text) {
    const digit = base58Map.get(char);
    if (digit == null) {
      throw new Error(`invalid base58 character ${char}`);
    }
    value = value * 58n + BigInt(digit);
  }
  const bytes: number[] = [];
  while (value > 0n) {
    bytes.push(Number(value & 0xffn));
    value >>= 8n;
  }
  for (const char of text) {
    if (char !== "1") {
      break;
    }
    bytes.push(0);
  }
  return new Uint8Array(bytes.reverse());
}

export function base58Encode(bytes: Uint8Array): string {
  let value = 0n;
  for (const byte of bytes) {
    value = (value << 8n) + BigInt(byte);
  }
  let text = "";
  while (value > 0n) {
    const mod = Number(value % 58n);
    text = base58Alphabet[mod] + text;
    value /= 58n;
  }
  for (const byte of bytes) {
    if (byte !== 0) {
      break;
    }
    text = "1" + text;
  }
  return text || "1";
}

export function base64Decode(text: string): Uint8Array {
  const binary = atob(text);
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) {
    out[i] = binary.charCodeAt(i);
  }
  return out;
}

function base64URLDecode(text: string): Uint8Array {
  const padded = text
    .replace(/-/g, "+")
    .replace(/_/g, "/")
    .padEnd(Math.ceil(text.length / 4) * 4, "=");
  return base64Decode(padded);
}

function base64URLEncode(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

function expectKeyBytes(bytes: Uint8Array, name: string): Uint8Array {
  if (bytes.byteLength !== 32) {
    throw new Error(`invalid ${name} length: ${bytes.byteLength}`);
  }
  return bytes;
}

function concatBytes(parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((sum, part) => sum + part.length, 0));
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

function arrayBufferFromBytes(data: Uint8Array): ArrayBuffer {
  const out = new ArrayBuffer(data.byteLength);
  new Uint8Array(out).set(data);
  return out;
}
